src/QirRuntime/lib/Tracer/README.md

```diff
@@ -0,0 +1,177 @@
+ is allocated and released. Start time of a gate cannot be less than allocation time of any of the qubits the gate uses.
+ If two gates or measurements use the same qubit, one of the gates must have start time greater than or equal to the end
+ time of the other.
+
+A sequentially executed quantum program can be assigned a trivial time function, when all quantum operations have
+ duration of 1 and unique start times, ordered to match the flow of the program. Layering compresses the timeline by
+ assuming that some operations might be executed simultaneously while allowing for different operations to have various
+ durations.
+
+_Definition_: Provided a valid _time_ function for the program a ___layer of duration N at time T, denoted as L(T,N),___
+ is a subset of operations in the program such that all of these operations have start time greater or equal _T_ and
+ finish time less than _T + N_. The program is ___layered___ if all gates in it are partitioned into layers, that don't
+ overlap in time. The union of all qubits that are involved in operations of a given layer, will be denoted _Qubits(T,N)_.
+
+A sequential program can be trivially layered such that each layer contains exactly one operation. Notice, that the
+ definition of layer doesn't require the gates to be executed _in parallel_. For example, all gates in a fully sequential
+ program can be also placed into a single layer L(0, infinity). Some gates might be considered to be very cheap and take
+ zero time to execute, those gates can be added to a layer even if they act on the same qubit another gate in this layer
+ is acting on and have to be executed sequentially within the layer.
+
+_Definition_: A ___global barrier___ is any operation that acts on _all_ currently allocated qubits. We'll provide means
+ for the clients to inject global barriers equivalent to an identity operator. This will allow the clients to enforce a
+ particular layering structure (because no later operation can sink below the barrier).
+
+The global barriers will be implemented as calls to `__quantum__qis__global_barrier` function. No additional support from
+ the compiler should be needed, as the user can define their own intrinsic to represent the barrier and map it to the above
+ runtime function via targets.qs file. The user can choose duration of a barrier which would affect start time of the
+ following layers but no operations will be added to a barrier, independent of its width.
+
+### The Resource Tracer's Layering Algorithm ###
+
+As the tracer is executing a sequential quantum program, it will compute a time function and corresponding layering
+ using the _conceptual_ algorithm, described below (aka "tetris algorithm"). The actual implementation of layering might
+ be done differently, as long as the resulting layering is the same as if running the conceptual algorithm.
+
+1. The tracer will have a setting for preferred layer duration: P.

```

Comment:
By this moment would be nice to understand why the tracer may need a setting for preferred layer duration.

Reply:
How else would you calculate the layers?